### PR TITLE
[[Bug 20326]] "Ask file" on Mac broken

### PR DESCRIPTION
**Pull request by bwmilby**
Whenever a save dialog should be presented, it failed silently.  If attempting to open a code editor after that it would crash LiveCode.  When looking at the compile logs, I noticed some potential errors and warnings.

Lines 359, 629-631 corrections were just to eliminate compiler warnings.

Line 634-635 change allowed the save dialog box to appear.

### CODE REVIEW

File: engine/src/mac-dialog.mm [hunk 1]
```diff
@@ -626,12 +626,16 @@ void MCPlatformBeginFolderOrFileDialog(MCPlatformFileDialogKind p_kind, MCPlatfo
...
+        [(NSOpenPanel *)t_panel setCanChooseFiles: NO];
+        [(NSOpenPanel *)t_panel setCanChooseDirectories: YES];
+        [(NSOpenPanel *)t_panel setAllowsMultipleSelection: NO];
+
+        // MM-2012-03-01: [[ BUG 10046]] Make sure the "new folder" button is enabled for folder dialogs
+        [t_panel setCanCreateDirectories: YES];
     }
     
-    if ([t_panel respondsToSelector:@selector(isAccessoryViewDisclosed)])
+    if (p_kind != kMCPlatformFileDialogKindSave &&
```
Comment (montegoulding) [inline engine/src/mac-dialog.mm]: @bwmilby not really sure what's up here unless apple decided to have NSSavePanel implement `isAccessoryViewDisclosed` just to throw an error...

Anyway... this all looks safe to rebase onto develop-8.1

Comment (montegoulding) [reply engine/src/mac-dialog.mm]: We should have both issues reported and notes for them though

Comment (bwmilby) [reply engine/src/mac-dialog.mm]: When I do the "make config-mac", there are 2 changes that I notice (between 8.1 and 9.0):
- SDK 10.8 --> 10.9
- target_arch=i386 --> i386 x86_64

I've added a new bug report and release notes.

Comment (runrevmark) [reply engine/src/mac-dialog.mm]: The accessory view problem I believe is related to SDK / Xcode combinations - I don't think it occurs in builds done from vulcan...

The folder issue is distinct and needs to be submitted in a separate PR against release-8.1.7 as that is a regression against a fix made in that release.

Comment (bwmilby) [reply engine/src/mac-dialog.mm]: Should the compiler warning corrections be moved to the 8.1.7 patch too?  That would leave this PR as just adjusting the accessory view code (either as I did already or moving it into the else clause which would make it match where it was logically before the change that introduced the issue).

Comment (runrevmark) [reply engine/src/mac-dialog.mm]: I think just fix the folder bug in release-8.1.7 - then do the rest in the fix to solve the problem the PR was originally for :)

File: engine/src/mac-dialog.mm [hunk 2]
```diff
@@ -626,12 +626,13 @@ void MCPlatformBeginFolderOrFileDialog(MCPlatformFileDialogKind p_kind, MCPlatfo
...
+        [(NSOpenPanel *)t_panel setCanChooseDirectories: YES];
+        [(NSOpenPanel *)t_panel setAllowsMultipleSelection: NO];
     }
     
-    if ([t_panel respondsToSelector:@selector(isAccessoryViewDisclosed)])
+    if (p_kind != kMCPlatformFileDialogKindSave && 
+        [t_panel respondsToSelector:@selector(isAccessoryViewDisclosed)])
     {
         // show accessory view when dialog opens
         [t_panel setAccessoryViewDisclosed: YES];
```
Comment (bwmilby) [inline engine/src/mac-dialog.mm]: This change does eliminate the eventual crash and allow the save dialog to appear, but does not eliminate the compiler warning.  I added a type cast and recompiled and still see the following:
```
/Users/milby/projects/GitHub/livecode/engine/src/mac-dialog.mm:638:33: warning: 
      instance method '-setAccessoryViewDisclosed:' not found (return type
      defaults to 'id') [-Wobjc-method-access]
        [(NSOpenPanel *)t_panel setAccessoryViewDisclosed: YES];
                                ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/milby/projects/GitHub/livecode/engine/src/mac-dialog.mm:17:
In file included from /Applications/Xcode-Dev/Xcode_8_3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/Cocoa.framework/Headers/Cocoa.h:13:
In file included from /Applications/Xcode-Dev/Xcode_8_3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/AppKit.framework/Headers/AppKit.h:76:
/Applications/Xcode-Dev/Xcode_8_3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSOpenPanel.h:13:12: note: 
      receiver is instance of class declared here
@interface NSOpenPanel : NSSavePanel {
           ^
2 warnings generated.
```
(the other warning is a redefine of nil, MacTypes.h defines it to be NULL)

**Comment by bwmilby** [comment]
Not sure this still should have the "need-rebase" tag since the other issue was split out specifically for 8.1.7

**Comment by montegoulding** [comment]
@bwmilby could you have a look at the conflicts here

**Comment by bwmilby** [comment]
Actually will need a third option.

```
        // MM-2012-03-01: [[ BUG 10046]] Make sure the "new folder" button is enabled for folder dialogs
        [t_panel setCanCreateDirectories: YES];
```

The above code needs to be retained in the merged code.  Those 2 lines were a part of 20335.  See also 6d413f8.

**Comment by montegoulding** [comment]
@bwmilby yes that's why there's the conflict. You need to manually merge everything between the conflict markers. It looks simple enough to do in GitHub's web editor

**Comment by montegoulding** [comment]
@livecode-vulcan review ok 0524082

Thanks @bwmilby

**Comment by livecode-vulcan** [comment]
💙 review by @montegoulding ok 052408282aa5d58c517f429526179709d717ef40

**Comment by livecode-vulcan** [comment]
😎 test success 052408282aa5d58c517f429526179709d717ef40

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/795)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/811)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/785)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/729)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/732)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/789)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/811)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/794)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/811)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/793)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/804)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/825)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/789)

**Pull request merged**